### PR TITLE
pkp/pkp-lib#1086 Version control for OJS3

### PR DESCRIPTION
**Pull request by felixgruender**

### CODE REVIEW

**Comment by felixgruender** [comment]
See also:
https://github.com/pkp/ojs/pull/651

File: classes/submission/PKPSubmissionMetadataFormImplementation.inc.php [hunk 1]
```diff
@@ -105,6 +105,11 @@ function initData($submission) {
 			// include all submission metadata fields for submissions
 			$this->_parentForm->setData('submissionSettings', array('all' => true));
 			$this->_parentForm->setData('abstractsRequired', $this->_getAbstractsRequired($submission));
+			
+			$submissionDao = Application::getSubmissionDAO();
+			$submissionSettingsRevisions = $submissionDao->getSubmissionRevisions($submission->getId(), true, true, 'DESC');
```
Comment (asmecher) [inline classes/submission/PKPSubmissionMetadataFormImplementation.inc.php]: You can use the `SORT_DIRECTION_...` constants defined in `DAO.inc.php` instead of ASC/DESC strings.

File: classes/submission/PKPSubmissionMetadataFormImplementation.inc.php [hunk 2]
```diff
@@ -147,6 +152,25 @@ function execute($submission, $request) {
...
+			$publicationTabFieldNames = $submissionDao->getAdditionalFieldNames();
+			
+			foreach($publicationTabFieldNames as $publicationTabFieldName) {
+				$submission->setData($publicationTabFieldName, $revisionSubmissionData->getData($publicationTabFieldName));
+			}
+		} else {
+			$revision = $submissionDao->getLatestRevisionId($submission->getId());
+		}
+
+		if ($request->getUserVar('saveAsRevision') && ($request->getUserVar('saveAsRevision') == 1)) {
```
Comment (asmecher) [inline classes/submission/PKPSubmissionMetadataFormImplementation.inc.php]: Wouldn't just...

    if ($request->getUserVar('saveAsRevision') ...

...be sufficient? Much more readable IMO.

File: classes/submission/SubmissionDAO.inc.php [hunk 3]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
+	function getSubmissionSettings($submission, $submissionRevision = null) {
+
+		if ($submissionRevision == null) {
+			$sql = "SELECT * FROM submission_settings WHERE submission_id = ?";
+			$params = array($submission->getId());
+		} else {
+			$sql = "SELECT * FROM submission_settings WHERE submission_id = ? AND submission_revision = ?";
+			$params = array($submission->getId(), $submissionRevision);
+		}
+		$result = $this->retrieve($sql, $params);
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: Consider writing this block as a single SQL statement, e.g.:

    $params = array((int) $submission->getId());
    if ($submissionRevision) $params[] = (int) $submissionRevision;
    $result = $this->retrieve(
        'SELECT * FROM submission_settings WHERE submission_id = ?' .
        ($submissionRevision?' AND submission_revision = ?':''), 
        $params
    );

(Also note the added int cast for submission revision.)

This duplicates less code and if we later had to add another potential condition it would scale better.

File: classes/submission/SubmissionDAO.inc.php [hunk 4]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
 
 		return $submission;
 	}
+	
+	/**
+	 * Get contents of the submission_settings table
+	 * @param $submission Submission object
+	 * @param $submissionRevision int optional
+	 */
+	function getSubmissionSettings($submission, $submissionRevision = null) {
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: The `get` prefix suggests it's going to return a value, but it loads it into the `$submission` instead. Suggest calling the function `loadSubmissionSettings` instead?

Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: (I wonder if we're going to have to scale out revisions beyond just the submissions table, e.g. to include author names, galley information, etc. In that case it might be worth making this an optional feature of DAO::getDataObjectSettings.)

File: classes/submission/SubmissionDAO.inc.php [hunk 5]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
+	 * 1) Retrieve only the previous revisions, not the current one
+	 * 2) Get only the id's; optionally the title can be retrieved, too
+	 * 3) Default sorting order is ASC
+	 * @param $submissionId int
+	 * @param $showAll boolean
+	 * @param $withTitle boolean
+	 * $param $order string
+	 * @return array
+	 */
+	function getSubmissionRevisions($submissionId, $showAll = false, $withTitle = false, $order = 'ASC') {
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: (Use constant for sort order; also probably elsewhere)

File: classes/submission/SubmissionDAO.inc.php [hunk 6]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
+		$result->Close();
+	}
+	
+	/**
+	 * Get all revisions for a submission; default settings: 
+	 * 1) Retrieve only the previous revisions, not the current one
+	 * 2) Get only the id's; optionally the title can be retrieved, too
+	 * 3) Default sorting order is ASC
+	 * @param $submissionId int
+	 * @param $showAll boolean
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: These parameters could use some inline documentation, e.g.:

    * @param $showAll boolean false to exclude current revision; true to fetch all revisions including current

File: classes/submission/SubmissionDAO.inc.php [hunk 7]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
+		if ($withTitle == false) {
+			$sql = 'SELECT DISTINCT submission_revision FROM submission_settings WHERE submission_id = ? ORDER BY submission_revision ' . $order;
+			$params = array($submissionId);
+		} else {
+			$sql = 'SELECT DISTINCT submission_revision, setting_value FROM submission_settings 
+				    WHERE submission_id = ? AND locale = ? AND setting_name = ? ORDER BY submission_revision ' . $order;
+			$params = array($submissionId, AppLocale::getPrimaryLocale(), 'title');
+		}
+		
+		$result = $this->retrieve($sql, $params);
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: As suggested above, consider writing this query out once with ternary operators for optional parts.

File: classes/submission/SubmissionDAO.inc.php [hunk 8]
```diff
@@ -111,10 +112,104 @@ function _fromRow($row) {
...
+		return $submissionRevisions;
+	}
+	
+	/**
+	 * Get the current revision id for a submission
+	 * @param $submissionId int
+	 * @param $journalId int
+	 * @return int
+	 */
+	function getLatestRevisionId($submissionId, $journalId = null) {
```
Comment (asmecher) [inline classes/submission/SubmissionDAO.inc.php]: This is in pkp-lib so shouldn't refer to OJS-specific aspects (e.g. journalId)

File: templates/controllers/wizard/fileUpload/form/submissionFileMetadataForm.tpl [hunk 9]
```diff
@@ -30,6 +30,25 @@
 		{fbvFormSection title="submission.upload.noteToAccompanyFile"}
 			{fbvElement type="textarea" id="note" height=$fbvStyles.height.SHORT}
 		{/fbvFormSection}
+		{if $submissionSettingsRevisions}
+			<label for="submissionSettingsRevision">{translate key="submission.upload.assignMetadata"}:</label>
+			<select name="submissionSettingsRevision">
```
Comment (asmecher) [inline templates/controllers/wizard/fileUpload/form/submissionFileMetadataForm.tpl]: Suggest writing this with `{fbvElement type="select" ...}` rather than a custom `<select>` element. Not that we're totally happy with this approach but it is consistent with the rest of the codebase.

File: templates/controllers/wizard/fileUpload/form/supplementaryFileMetadataForm.tpl [hunk 10]
```diff
@@ -38,6 +38,25 @@
 			{fbvElement label="common.date" inline=true size=$fbvStyles.size.SMALL type="text" id="dateCreated" value=$submissionFile->getDateCreated(null) class="datepicker"}
 			{fbvElement label="common.language" inline=true size=$fbvStyles.size.SMALL type="text" id="language" value=$submissionFile->getLanguage() maxlength="255"}
 		{/fbvFormSection}
+		{if $submissionSettingsRevisions}
+			<label for="submissionSettingsRevision">{translate key="submission.upload.assignMetadata"}:</label>
+			<select name="submissionSettingsRevision">
```
Comment (asmecher) [inline templates/controllers/wizard/fileUpload/form/supplementaryFileMetadataForm.tpl]: (Use fbvElement)

File: templates/submission/submissionMetadataFormTitleFields.tpl [hunk 11]
```diff
@@ -8,7 +8,15 @@
...
  * submission metadata.
  *}
+{if $formParams.revision && ($formParams.revision < $latestRevisionId)}
+	{assign var=rich value=null}
+{else}
+	{assign var=rich value="extended"}
+{/if}
+
 {fbvFormArea id="generalInformation" title="submission.submit.titleAndSummary"}
+	{fbvElement type="hidden" name="submissionRevision" id="submissionRevision" value=$formParams.revision|escape}
```
Comment (asmecher) [inline templates/submission/submissionMetadataFormTitleFields.tpl]: When using fbvElement, escaping is taken care of.

File: templates/submission/submissionMetadataFormTitleFields.tpl [hunk 12]
```diff
@@ -21,6 +29,6 @@
 	{/fbvFormSection}
 	{fbvFormSection description="common.subtitle.tip"}{/fbvFormSection}
 	{fbvFormSection title="common.abstract" for="abstract" required=$abstractsRequired}
-		{fbvElement type="textarea" multilingual=true name="abstract" id="abstract" value=$abstract rich="extended" readonly=$readOnly}
+		{fbvElement type="textarea" multilingual=true name="abstract" id="abstract" value=$abstract rich=$rich readonly=$readOnly}
```
Comment (asmecher) [inline templates/submission/submissionMetadataFormTitleFields.tpl]: Is this change intentional?

File: xml/oai2.xsl [hunk 13]
```diff
@@ -598,7 +598,7 @@ p.intro {
   <xsl:choose>
     <xsl:when test='starts-with(.,"http" )'>
       <xsl:choose>
-        <xsl:when test='string-length(.) &gt; 50'>
+        <xsl:when test='string-length(.) &gt; 80'>
```
Comment (asmecher) [inline xml/oai2.xsl]: Is this change intentional? It's 3rd-party code, which we try not to distribute modified copies of if possible.

File: classes/submission/SubmissionFileDAODelegate.inc.php [hunk 14]
```diff
@@ -359,6 +362,18 @@ function _deleteDependentObjects($submissionFile) {
 				ASSOC_TYPE_SUBMISSION_FILE, $submissionFile->getFileIdAndRevision()
 		);
 	}
+	
+	/**
+	 * @copydoc DAO::getAdditionalFieldNames()
+	 */
+	function getAdditionalFieldNames() {
```
Comment (asmecher) [inline classes/submission/SubmissionFileDAODelegate.inc.php]: I think I'd prefer to use a new column in the relevant tables rather than storing this as a setting. It will be a core feature, and there are likely to be performance implications that require some optimization, and that'll be easier if it's a full-fledged column.

File: templates/submission/submissionMetadataFormTitleFields.tpl (hunk 12 above)
Comment (felixgruender) [inline templates/submission/submissionMetadataFormTitleFields.tpl]: Yes, that was intentional. For the old revisions, I deactivate all input fields - and I realized that deactivating a textarea is only possible when it's not in the "rich" mode. That means that I have to turn off the "rich" mode first when I want to make it read-only.

Comment (asmecher) [inline templates/submission/submissionMetadataFormTitleFields.tpl]: I see. That'll be a problem if there is rich content, because turning off `rich` will result in HTML tags etc. being escaped and displayed to the reader literally.

I can think of two approaches, both best solved in the form builder vocab code:
- Configure TinyMCE to behave read-only as expected when `readonly` is set, or
- have the form builder vocab present a `<div>` element or something, with the contents of the field inside it, and not use TinyMCE at all for cases where `readonly` is set

**Comment by asmecher** [comment]
Superseded by https://github.com/pkp/pkp-lib/pull/1471
